### DISCUSSION TITLE
Distance from Julia?

### DISCUSSION THREAD

**Discussion by Quafadas** [markehammons/slinc#130] [Ideas]
A curiosity: a guess on how possible it might to port the concepts here to allow Julia interop? 

If forced to guess: could it be functionality which could live peacefully in this library, or would it rather be a standalone effort?

**Comment by markehammons** [comment]
I'm not familiar with the ins and outs of julia, but looking at its documentation, this would be the way:
https://docs.julialang.org/en/v1/manual/embedding/

Looking at this project: https://github.com/matthijscox/embedjuliainc it looks like aside from invoking the julia init code and shutdown code, it should be possible to bind to the julia files by loading them and binding to them like C. I will have to investigate, but it doesn't look like this would be very hard, assuming one can easily generate so files for julia.

**Comment by Quafadas** [comment]
Wow ... thanks for writing back. It sounds like the barrier _may_ not be that high, with the caveat that we must be able to generate "so" files for Julia. I had a poke around, and it would appear that this Julia package appears to explicitly do this;
https://julialang.github.io/PackageCompiler.jl/dev/libs.html

It's documentation says

> (These will have a .so extension on Linux, and a .dll extension on Windows. There may also be other files in the same directory, depending on your operating system and version of Julia.)

So based on all this, I'd certainly willing to put in some of my own time on a POC that sets out to call a trivial Julia library from Scala. 

I tried to read the docs for slinc library, but from what I could tell, the link to docs from the readme is broken... would you expect them to still be published?

Finally to calibrate my expectations: 

> it doesn't look like this would be very hard

Should I read this to mean this may be possible for a consumer of this library as is, carefully following instructions (me)? Or rather that it may be possible for a skilled developer to add it as a feature?

**Reply by markehammons** [reply]
I'm fairly certain that this library, as is, could be used to use julia functions from Scala. I'd suggest for your PoC that you use one of the latest versions of Slinc (0.3.0 should be available on Monday), and try calling functions from a julia .so. From there, you may want to add code to the project that eases interaction with julia, but before you test it's kind of hard to know what the pain points will be.

**Reply by Quafadas** [reply]
Very exciting - thanks again. I'll put some time aside in the coming weeks to dig in.

**Comment by Quafadas** [comment]
https://github.com/Quafadas/slincTest

barfs with 

```
WARNING: failed to initialize stack walk info
ERROR: could not load library "C:\temp\superJlibCompiled\lib\julia\sys.dll"
The specified module could not be found.
```

I interpret this to mean that some part of the Julia C interface isn't correctly linked. 

https://discourse.julialang.org/t/embedding-julia-in-scala/97347

I can solve that, apparently, by simply copying the file out of the vanilla Julia install to the right place... but then I get this; 

```
Exception in thread "main" java.lang.ExceptionInInitializerError
        at calc.main(slinc.scala:29)
Caused by: java.util.NoSuchElementException: None.get
        at scala.None$.get(Option.scala:627)
        at scala.None$.get(Option.scala:626)
        at fr.hammons.slinc.modules.FSetModule17$package$fsetModule17$.$anonfun$1(FSetModule17.scala:31)
        at scala.collection.immutable.List.map(List.scala:250)
        at fr.hammons.slinc.modules.FSetModule17$package$fsetModule17$.getBacking(FSetModule17.scala:53)
        at fr.hammons.slinc.FSet.get(FSet.scala:26)
        at fr.hammons.slinc.FSet.get$(FSet.scala:13)
        at JuliaLib$$anon$6.get(slinc.scala:45)
        at slinc$package$.<clinit>(slinc.scala:26)
```
When trying to call a function from the package. I interpret that again, to mean that I haven't got the linking right. Hmmmm.

Which I think is because I didn't have the right dll / method name combination
```
Please submit a bug report with steps to reproduce this fault, and any error messages that follow (in their entirety). Thanks.
Exception: EXCEPTION_ACCESS_VIOLATION at 0x0 -- unknown function (ip: 0000000000000000)
in expression starting at none:0
unknown function (ip: 0000000000000000)
Allocations: 2907 (Pool: 2896; Big: 11); GC: 0
```

**Reply by Quafadas** [reply]
So it definitely works to call Julia, almost trivially from this library, and the code suggested in the "embedding" guide works. 

So I think this library calls Julia's C interface correctly. Which is kind of aweseome, 
```
  juliaLib.jl_init()
  
  Scope.confined {
    println("Code supplied to Julia directly")    
    val t = Ptr.copy("print(sqrt(2.0))")
    juliaLib.jl_eval_string(t);
  }
  
  println("\n Now let's call a pre-compiled library")
  println(superJ.increment32(1))
```
Outputs
```
WARNING: failed to initialize stack walk info
Code supplied to Julia directly
1.4142135623730951
 Now let's call a pre-compiled library

Please submit a bug report with steps to reproduce this fault, and any error messages that follow (in their entirety). Thanks.
Exception: EXCEPTION_ACCESS_VIOLATION at 0x0 -- unknown function (ip: 0000000000000000)
in expression starting at none:0
unknown function (ip: 0000000000000000)
Allocations: 2907 (Pool: 2896; Big: 11); GC: 0
```
I’m increasingly convinced this error message is something to do with linking / compiling Julia libraries rather than slinc's interface to them . 

I suspect that my next step, should be to write a program to call this library from C itself (which Ihave no idea how to do), and see if that works.

**Reply by Quafadas** [reply]
https://github.com/JuliaLang/PackageCompiler.jl/issues/803

If I get the chance, I'll try dropping down to C, and seeing if I can succeed from there, and if that success helps my understanding of what I might be doing wrong.

I suspect, that the C compiler establishes some links between the Julia output and other DLLs, and that secondary set of links is currently missing from my naive slic call to the Julia dll, which was produced by PackageCompiler.jl. 

I'm now out of my depth. In conclusion; 
- I was able to prove that this library successfully calls the Julia C interface (quodos!). 
- I could (apparently) compile a (trivial) Julia library, but calling it via it's C interface, yielded a low level exception (above). I do not believe I have the skills to solve this.

**Reply by markehammons** [reply]
@Quafadas I think your slincExample repository is private, cause I cannot access it. I'd like to look at it and see what I can do.

**Reply by Quafadas** [reply]
Huh, that was a mistake. It's now public. Whoops - thanks for telling me.

**Reply by Quafadas** [reply]
@markehammons 

I've had a dig into the static compiler package, and produced a "static" .so file. I can call this successfully from Julia itself, but not from scala, where I get a nasty looking memory exception. 

I'm trying to follow the example given  here; 
https://github.com/brenhinkeller/StaticTools.jl#compiled-sodylib-shared-libraries

I suspect, that the "shape" of the class I'm trying to feed the API call, might not be quite right somehow. If you had two mins, to see if you think there is something obviously "wrong" with what I'm attempting... I'd be very grateful.

https://github.com/Quafadas/slincTest/blob/0f58d0443ae0dc2902271578229f401842097ad7/src/main/scala/slinc.scala#L40-L41

No pressure to investigate, if you've already had enough of this idea! Log looks like this; 

```
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x00007f6854095ec4, pid=11618, tid=11709
#
# JRE version: OpenJDK Runtime Environment Microsoft-7209853 (17.0.6+10) (build 17.0.6+10-LTS)
# Java VM: OpenJDK 64-Bit Server VM Microsoft-7209853 (17.0.6+10-LTS, mixed mode, sharing, tiered, compressed oops, compressed class ptrs, g1 gc, linux-amd64)
# Problematic frame:
# C  [libmul.so+0x3ec4]  julia_mul_inplace+0x4

```

**Reply by Quafadas** [reply]
https://quafadas-literate-space-goldfish-g5qw954xj6f9qgp.github.dev/ is a codespace which _might_ help to investigate more closely.

**Reply by Quafadas** [reply]
https://github.com/Quafadas/slincTest/blob/master/hs_err_pid4342.log full log (I don't understand anything in there)

**Reply by Quafadas** [reply]
@markehammons Apologies for the continued spam, but I think I now have a much more exciting result, having successfully called a statically compiled library from slinc. 

https://github.com/brenhinkeller/StaticTools.jl/issues/43#issuecomment-1511636729

On the other hand, it currently claims that 1 + 2 is infinite... could this be some impedance between CFloat and Julias Int64? It's definitely a demonstration, that this might be a useful direction of travel though?
